Crun 18/19

ayuranjan opened the pull request:
-  Implemented Vocabulary    ( initial code and test cases  was already pushed to different branch which was merged to bootstrap branch by @adhocmaster,  so not much change ) 
   - For duplicate words we just have to ignore them 
- Implemented Incremental Vocabulary 
   - For duplicate words we just have to throw error  
- Created test cases for the cases required

sdmuki commented on src/implementations/nlp/IncrementalVocabulary.ts:
```diff
@@ -1,12 +1,22 @@
+  * @description: adds words/ word roots to the vocabulary
+  * @throws Err if the word/ word root is already present in the vocabulary
+  * */
+  public addWord(word: Word | WordRoot): void {
```
we will need to change the return types to ResultAsync. No throwing of errors

sdmuki commented on src/implementations/nlp/Vocabulary.ts:
```diff
@@ -3,26 +3,60 @@ import { VocabularySize, Word, WordRoot, WordIndex, WordToIndex } from "crunchDB
         for (let i = 0; i < words.length; i++) {
-            this.wordToIndex[words[i] as string] = WordIndex(i);
-            this.indexToWord[i] = words[i];
+            if(! this.wordToIndex.hasOwnProperty(words[i])) // duplicate check 
```
maybe we should convert the wordToIndex to map types instead of object type. Check typescript map type

sdmuki commented on src/implementations/nlp/Vocabulary.ts:
```diff
@@ -3,26 +3,60 @@ import { VocabularySize, Word, WordRoot, WordIndex, WordToIndex } from "crunchDB
+    * @returns WordIndex
+    * @description: returns the index of the word/ word root
+    * */
     public getWordIndex(word: Word | WordRoot): WordIndex {
```
As I thought about it more, there will be words not existing in our vocabulary. So, this function should be allowed to return null.

ayuranjan replied:
Yes, that would be better. Will change the return type to (WordIndex | null)  in this case.

sdmuki commented on test/implementations/nlp/IncrementalVocabulary.test.ts:
```diff
@@ -0,0 +1,39 @@
+  it("should throw error for duplicate word", () => {
+    
+    try {
+        vocabulary.addWord(Word("apple"));
```
There will be almost no try catch. I will write a pattern to test with ResultAsync return type soon.

ayuranjan replied:
So, regarding this should I put this on hold as of now?

ayuranjan commented:
Changes done in the new commit : 

- converted  WordToIndex to map types instead of object type. 
- Changed the return type of IncrementalVocab's addWord() to ResultAsync and modified the Test Case accordingly.  Created a new Error i.e DuplicateWordError for the case when the word you are adding is already present. 
- Changed the return type of  getWordIndex() to WordIndex | null, Since there will be words not existing in our vocabulary and in such case this function will return null.  Added Test Case as well for this condition.

SnickerChar commented on src/implementations/nlp/IncrementalVocabulary.ts:
```diff
@@ -1,12 +1,25 @@
+    else{
+      this.wordToIndex.set(word, WordIndex(this.wordToIndex.size));
+      this.indexToWord.push(word); 
+      return okAsync(void 0);
```
You can `return okAsync(undefined)` to meet a "void" return type, that's what we do everywhere else.

SnickerChar commented on test/implementations/nlp/IncrementalVocabulary.test.ts:
```diff
@@ -0,0 +1,25 @@
+import { IncrementalVocabulary } from  "crunchDB/implementations";
+import {  Word, WordRoot } from "crunchDB/objects";
+describe("IncrementalVocabulary", () => {
+  const words: (Word | WordRoot)[] = [Word("apple"), Word("banana"), Word("orange")];
```
Don't use a shared object between unit tests. Instead, create a class IncrementalVocabularyMocks that has the values you need. Then, in each individual test, do a `const mocks = new IncrementalVocabularyMocks();` The reason why is that is prevents coupling between the tests and allows them to be run in any order, which is important for unit tests.

ayuranjan replied:
I have used beforeEach() hooks which apply to every test in a file and since the initial setup is done inside that hook so changing the order of test won't impact any other test.
